KerbalStuff/kerbdown.py

```diff
@@ -1,69 +1,75 @@
 import urllib.parse
 from urllib.parse import parse_qs, urlparse
-from typing import Dict, Any, Match
+from typing import Dict, Any, Match, Tuple
 
 from markdown import Markdown
 from markdown.extensions import Extension
-from markdown.inlinepatterns import Pattern
+from markdown.inlinepatterns import InlineProcessor
 from markdown.util import etree
 
-EMBED_RE = r'\[\[(?P<url>.+?)\]\]'
 
+class EmbedInlineProcessor(InlineProcessor):
+    # Don't worry about re.compiling this, markdown.inlinepatterns.Pattern.__init__ does that for us
+    EMBED_RE = r'\[\[(?P<url>.+?)\]\]'
 
-def embed_youtube(link: urllib.parse.ParseResult) -> etree.Element:
-    q = parse_qs(link.query)
-    v = q['v'][0]
-    el = etree.Element('iframe')
-    el.set('width', '100%')
-    el.set('height', '600')
-    el.set('frameborder', '0')
-    el.set('allowfullscreen', '')
-    el.set('src', '//www.youtube-nocookie.com/embed/' + v + '?rel=0')
-    return el
+    # Prefixes of the iframe src attributes we generate
+    YOUTUBE_SRC_PREFIX = '//www.youtube-nocookie.com/embed/'
+    IMGUR_SRC_PREFIX = '//imgur.com/a/'
+    IFRAME_SRC_PREFIXES = [YOUTUBE_SRC_PREFIX, IMGUR_SRC_PREFIX]
 
+    # Other iframe attributes we generate
+    IFRAME_ATTRIBS = ['width', 'height', 'frameborder', 'allowfullscreen']
 
-def embed_imgur(link: urllib.parse.ParseResult) -> etree.Element:
-    a = link.path.split('/')[2]
-    el = etree.Element('iframe')
-    el.set('width', '100%')
-    el.set('height', '550')
-    el.set('frameborder', '0')
-    el.set('allowfullscreen', '')
-    el.set('src', '//imgur.com/a/' + a + '/embed')
-    return el
-
-
-class EmbedPattern(Pattern):
-    def __init__(self, pattern: str, m: Markdown, configs: Dict[str, Any]) -> None:
-        super(EmbedPattern, self).__init__(pattern, m)
+    def __init__(self, md: Markdown, configs: Dict[str, Any]) -> None:
+        super().__init__(self.EMBED_RE, md)
         self.config = configs
 
-    def handleMatch(self, m: Match[str]) -> etree.Element:
+    def handleMatch(self, m: Match[str], data: str) -> Tuple[etree.Element, int, int]:  # type: ignore[override]
         d = m.groupdict()
         url = d.get('url')
         if not url:
             el = etree.Element('span')
             el.text = "[[]]"
-            return el
+            return el, m.start(0), m.end(0)
         try:
             link = urlparse(url)
             host = link.hostname
         except:
             el = etree.Element('span')
             el.text = "[[" + url + "]]"
-            return el
+            return el, m.start(0), m.end(0)
         el = None
         try:
             if host == 'youtube.com' or host == 'www.youtube.com':
-                el = embed_youtube(link)
+                el = self._embed_youtube(link)
```

Comment:
Nowadays YouTube generates "share links" in this form: `https://youtu.be/dQw4w9WgXcQ`
Should we support them as well? It needs some changes to `_embed_youtube()` though, as the video id is no longer in the `v` query string, but the path.

Reply:
OK, done.